man/kclangctest.1
.TH "KCLANGCTEST" 1 "2011-02-06" "Man Page" "Kyoto Cabinet"

.SH NAME
kclangctest \- command line interface to test the C language binding

.SH DESCRIPTION
.PP
The command `\fBkclangctest\fR' is a utility for facility test and performance test of the C language binding.  This command is used in the following format.  `\fIpath\fR' specifies the path of a database file.  `\fIrnum\fR' specifies the number of iterations.
.PP
.RS
.br
\fBkcpolytest order \fR[\fB\-rnd\fR]\fB \fR[\fB\-etc\fR]\fB \fR[\fB\-tran\fR]\fB \fR[\fB\-oat\fR|\fB\-onl\fR|\fB\-onl\fR|\fB\-otl\fR|\fB\-onr\fR]\fB \fIpath\fB \fIrnum\fB\fR
.RS
Performs in\-order tests.
.RE
.RE
.PP
Options feature the following.
.PP
.RS
\fB\-rnd\fR : performs random test.
.br
\fB\-etc\fR : performs miscellaneous operations.
.br
\fB\-tran\fR : performs transaction.
.br
\fB\-oat\fR : opens the database with the auto transaction option.
.br
\fB\-oas\fR : opens the database with the auto synchronization option.
.br
\fB\-onl\fR : opens the database with the no locking option.
.br
\fB\-otl\fR : opens the database with the try locking option.
.br
\fB\-onr\fR : opens the database with the no auto repair option.
.br
.RE
.PP
This command returns 0 on success, another on failure.

.SH SEE ALSO
.PP
.BR kcpolymgr (1),
.BR kcpolytest (1)


ChangeLog
2011-02-12  Mikio Hirabayashi  <info@fallabs.com>

	* kchashdb.h (HashDB): the type of record locking objects was changed.

	* kcdirdb.h (DirDB): the type of record locking objects was changed.

	* kcplantdb.h (PlantDB): the type of page locking objects was changed.

	- Release: 1.2.43

2011-02-10  Mikio Hirabayashi  <info@fallabs.com>

	* kcdb.h (DB::Cursor::get_key, DB::Cursor::get_value): signature was modified.

	* kcdb.h (DB::Cursor::get_pair, DB::Cursor::get): renamed.

	* kcdb.h (DB::get): signature was modified.

	- Release: 1.2.42

2011-02-07  Mikio Hirabayashi  <info@fallabs.com>

	* kccommon.h: macros for integer range were added.

	* kchashdb.h (HashDB::begin_transaction): behavior of locking was modified.

	* kcdirdb.h (DirDB::begin_transaction): behavior of locking was modified.

	- Release: 1.2.41

2011-02-05  Mikio Hirabayashi  <info@fallabs.com>

	* kccompare.h (LexicalDescendingComparator, LexicalDescendingComparator): new classes.

	* kccompare.h (LEXICALCOMP, DECIMALCOMP): the type was changed to pointer.

	* kccompress.h (ZLIBRAWCOMP): the type was changed to pointer.

	* kcplantdb.h (PlantDB::load_meta): a bug related to custom comparator was fixed.

	- Release: 1.2.40

2011-02-01  Mikio Hirabayashi  <info@fallabs.com>

	* kcfile.cc (File::open): checking type type was added.

	* kcplantdb.h (Plantdb::reorganize_file): recovery condition was modified.

	* kcpolymgr.cc (proccheck): a bug related to path handling was fixed.

	- Release: 1.2.39

2011-01-30  Mikio Hirabayashi  <info@fallabs.com>

	* kcthread.cc: threshold of sleep locking was modified.

	* kcplantdb.h (PlantDB::accept): a bug related to race condition was fixed.

	* kcplantdb.h (PlantDB::save_leaf_node): locking mode was modified.

	- Release: 1.2.38

2011-01-25  Mikio Hirabayashi  <info@fallabs.com>

	* kcutil.h (strvecdump, strvecload, strmapdump, strmapload): new functions.

	- Release: 1.2.37

2011-01-21  Mikio Hirabayashi  <info@fallabs.com>

	* kcutil.cc (getsysinfo): a bug related to memory stats was fixed.

	* kcthread.cc (Thread::chill): new function.

	* kcthread.cc (ScopedMutex, ScopedSpinLock): refactored using primitives.

	* kcthread.cc (ScopedRWLock, ScopedSpinRWLock): refactored using primitives.

	- Release: 1.2.36

2011-01-11  Mikio Hirabayashi  <info@fallabs.com>

	* kcutil.cc: the order of header inclusion was modified.

	* kcthread.cc (SpinRWLock): space efficiency was improved.

	- Release: 1.2.35

2011-01-09  Mikio Hirabayashi  <info@fallabs.com>

	* kcplantdb.h (PlantDB::accept, PlantDB::accept_bulk): behavior of async was modified.

	- Release: 1.2.34

2011-01-05  Mikio Hirabayashi  <info@fallabs.com>

	* kcfile.cc (File::open): handling the WAL file was modified.

	* kchashdb.h (HashDB::clear): handling the open flag was added.

	* kcplantdb.h (PlantDB::clear): handling the recovery flag was modified.

	- Release: 1.2.33

2011-01-03  Mikio Hirabayashi  <info@fallabs.com>

	* kcfile.cc (File::read_fast): error check was added.

	* kclangc.cc (kcdbacceptbulk, kcdbsetbulk, kcdbremovebulk, kcdbgetbulk): new functions.

	- Release: 1.2.32

2010-12-30  Mikio Hirabayashi  <info@fallabs.com>

	* kcdb.h (BasicDB::accept_bulk): new function.

	* kcdb.h (BasicDB::set_bulk, BasicDB::remove_bulk, BasicDB::get_bulk): new functions.

	* kcthread.h (SlottedMutex::unlock_all): the order became forward.

	* kcpolymgr.cc (procsetbulk, procremovebulk, procgetbulk): new functions.

	- Release: 1.2.31

2010-12-19  Mikio Hirabayashi  <info@fallabs.com>

	* kcpolydb.h (PolyDB::set_internal_db): new function.

	- Release: 1.2.30

2010-12-11  Mikio Hirabayashi  <info@fallabs.com>

	* kccommon.h (modfl): new function for portability to cygwin.

	* kcthread.cc (Mutex::lock_try): cygwin was added to work around cases.

	* cmdcommon.h (oprintf): new function instead of "iprintf".

	- Release: 1.2.29

2010-12-10  Mikio Hirabayashi  <info@fallabs.com>

	* ktutil.h (atoin, atofn): new functions.

	- Release: 1.2.28

2010-11-30  Mikio Hirabayashi  <info@fallabs.com>

	* kcfile.cc (File::write, File::read): warnings on 32-bit environment were cleared.

	* kcfile.cc (File::write_file, File::make_directory): modified for Win32 virus checkers.

	- Release: 1.2.27

2010-11-28  Mikio Hirabayashi  <info@fallabs.com>

	* ktutil.h (vstrprintf): solved a compilation warning.

	* kcdbext.h (MapReduce::~MapReduce): modified as a virtual function.

	- Release: 1.2.26

2010-11-17  Mikio Hirabayashi  <info@fallabs.com>

	* kcdb.h (BasicDB::MetaTrigger): new class.

	* kcdb.h (BasicDB::tune_trigger): new function.

	* kcpolydb.h (PolyDB::StreamMetaTrigger): new class.

	- Release: 1.2.25

2010-11-10  Mikio Hirabayashi  <info@fallabs.com>

	* kcstashdb.h, kcstashdb.o: new files.

	- Release: 1.2.24

2010-11-06  Mikio Hirabayashi  <info@fallabs.com>

	* kcutil.h (sizevarnum): new function.

	* kcmap.h (TinyHashMap): new class.

	* kcdbext.h (MapReduce::execute): the cache algorithm was modified.

	- Release: 1.2.23

2010-11-04  Mikio Hirabayashi  <info@fallabs.com>

	* kcutil.h (strsplit): new function.

	* kcpolydb.h (PolyDB::match_prefix): performance of the prototype tree was improved.

	- Release: 1.2.22

2010-11-02  Mikio Hirabayashi  <info@fallabs.com>

	* kcdbext.h (MapReduce::Emitter::emit): a bug on 32-bit environment was fixed.

	* kcdbext.h (MapReduce::execute): the naming convention of databases was modified.

	* kcregex.cc (Regex::compile): a bug related to error handling was fixed.

	- Release: 1.2.21

2010-10-29  Mikio Hirabayashi  <info@fallabs.com>

	* kcdbext.h, kcdbext.cc: new files.

	* kcpolytest.cc (runmapred, procmapred): new functions.

	- Release: 1.2.20

2010-10-21  Mikio Hirabayashi  <info@fallabs.com>

	* kcpolydb.h (PolyDB::match_prefix, PolyDB::match_regex): new functions.

	* kclangc.h (kcdbmatchprefix, kcdbmatchregex): new functions.

	* kcregex.h: new file.

	- Release: 1.2.19

2010-10-15  Mikio Hirabayashi  <info@fallabs.com>


kclangctest.c
/*************************************************************************************************
 * The test cases of the C language binding
 *                                                               Copyright (C) 2009-2011 FAL Labs
 * This file is part of Kyoto Cabinet.
 * This program is free software: you can redistribute it and/or modify it under the terms of
 * the GNU General Public License as published by the Free Software Foundation, either version
 * 3 of the License, or any later version.
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
 * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the GNU General Public License for more details.
 * You should have received a copy of the GNU General Public License along with this program.
 * If not, see <http://www.gnu.org/licenses/>.
 *************************************************************************************************/


#include <kclangc.h>

#define RECBUFSIZ    64                  /* buffer size for a record */
#define RECBUFSIZL   1024                /* buffer size for a long record */
#if !defined(TRUE)
#define TRUE         1                   /* boolean true */
#endif
#if !defined(FALSE)
#define FALSE        0                   /* boolean false */
#endif

typedef struct {                         /* arguments of visitor */
  int64_t rnum;
  int32_t rnd;
  int64_t cnt;
  char rbuf[RECBUFSIZ];
} VISARG;


/* global variables */
const char* g_progname;                  /* program name */


/* function prototypes */
int main(int argc, char** argv);
static void usage(void);
static int64_t myrand(int64_t range);
static void iprintf(const char* format, ...);
static void iputchar(char c);
static void eprintf(const char* format, ...);
static void dberrprint(KCDB* db, int32_t line, const char* func);
const char* visitfull(const char* kbuf, size_t ksiz,
                      const char* vbuf, size_t vsiz, size_t* sp, void* opq);
static int32_t runorder(int argc, char** argv);
static int32_t procorder(const char* path, int64_t rnum, int32_t rnd, int32_t etc,
                         int32_t tran, int32_t oflags);


const char* myfull(const char* kbuf, size_t ksiz,
                   const char* vbuf, size_t vsiz, size_t* sp, void* opq) {

  fwrite(kbuf, 1, ksiz, stdout);
  putchar('\t');
  fwrite(vbuf, 1, vsiz, stdout);
  putchar('\n');

  return KCVISNOP;
}

void setval(KCSTR* str, const char* val) {
  str->buf = (char*)val;
  str->size = strlen(val);
}


/* main routine */
int main(int argc, char **argv) {
  int32_t i, rv;
  g_progname = argv[0];
  srand(time(NULL));
  if (argc < 2) usage();
  rv = 0;
  if (!strcmp(argv[1], "order")) {
    rv = runorder(argc, argv);
  } else {
    usage();
  }
  if (rv != 0) {
    iprintf("FAILED:");
    for (i = 0; i < argc; i++) {
      iprintf(" %s", argv[i]);
    }
    iprintf("\n\n");
  }
  return rv;
}


/* print the usage and exit */
static void usage() {
  eprintf("%s: test cases of the C binding of Kyoto Cabinet\n", g_progname);
  eprintf("\n");
  eprintf("usage:\n");
  eprintf("  %s order [-rnd] [-etc] [-tran] [-oat|-oas|-onl|-otl|-onr] path rnum\n",
          g_progname);
  eprintf("\n");
  exit(1);
}


/* get a random number */
static int64_t myrand(int64_t range) {
  uint64_t base, mask;
  if (range < 2) return 0;
  base = range * (rand() / (RAND_MAX + 1.0));
  mask = (uint64_t)rand() << 30;
  mask += (uint64_t)rand() >> 2;
  return (base ^ mask) % range;
}


/* print formatted error string and flush the buffer */
static void iprintf(const char* format, ...) {
  va_list ap;
  va_start(ap, format);
  vprintf(format, ap);
  va_end(ap);
  fflush(stdout);
}


/* print a character and flush the buffer */
static void iputchar(char c) {
  putchar(c);
  fflush(stdout);
}


/* print formatted error string and flush the buffer */
static void eprintf(const char* format, ...) {
  va_list ap;
  va_start(ap, format);
  vfprintf(stderr, format, ap);
  va_end(ap);
  fflush(stderr);
}


/* print error message of database */
static void dberrprint(KCDB* db, int32_t line, const char* func) {
  char* path;
  const char* emsg;
  int32_t ecode;
  path = kcdbpath(db);
  ecode = kcdbecode(db);
  emsg = kcdbemsg(db);
  iprintf("%s: %d: %s: %s: %d: %s: %s\n",
          g_progname, line, func, path ? path : "-", ecode, kcecodename(ecode), emsg);
  kcfree(path);
}


/* print members of database */
static void dbmetaprint(KCDB* db, int32_t verbose) {
  char* status, *rp;
  if (verbose) {
    status = kcdbstatus(db);
    if (status) {
      rp = status;
      while (*rp != '\0') {
        if (*rp == '\t') {
          printf(": ");
        } else {
          putchar(*rp);
        }
        rp++;
      }
      kcfree(status);
    }
  } else {
    iprintf("count: %ld\n", (long)kcdbcount(db));
    iprintf("size: %ld\n", (long)kcdbsize(db));
  }
}


/* visit a full record */
const char* visitfull(const char* kbuf, size_t ksiz,
                      const char* vbuf, size_t vsiz, size_t* sp, void* opq) {
  VISARG* arg;
  const char* rv;
  arg = opq;
  arg->cnt++;
  rv = KCVISNOP;
  switch (arg->rnd ? myrand(7) : arg->cnt % 7) {
    case 0: {
      rv = arg->rbuf;
      *sp = arg->rnd ? (size_t)myrand(sizeof(arg->rbuf)) :
        sizeof(arg->rbuf) / (arg->cnt % 5 + 1);
      break;
    }
    case 1: {
      rv = KCVISREMOVE;
      break;
    }
  }
  if (arg->rnum > 250 && arg->cnt % (arg->rnum / 250) == 0) {
    iputchar('.');
    if (arg->cnt == arg->rnum || arg->cnt % (arg->rnum / 10) == 0)
      iprintf(" (%08ld)\n", (long)arg->cnt);
  }
  return rv;
}


/* parse arguments of order command */
static int32_t runorder(int argc, char** argv) {
  int32_t argbrk = FALSE;
  const char* path, *rstr;
  int32_t rnd, etc, tran;
  int32_t i, mode, oflags, rnum;
  path = NULL;
  rstr = NULL;
  rnd = FALSE;
  etc = FALSE;
  mode = 0;
  tran = FALSE;
  oflags = 0;
  for (i = 2; i < argc; i++) {
    if (!argbrk && argv[i][0] == '-') {
      if (!strcmp(argv[i], "--")) {
        argbrk = TRUE;
      } else if (!strcmp(argv[i], "-rnd")) {
        rnd = TRUE;
      } else if (!strcmp(argv[i], "-etc")) {
        etc = TRUE;
      } else if (!strcmp(argv[i], "-tran")) {
        tran = TRUE;
      } else if (!strcmp(argv[i], "-oat")) {
        oflags |= KCOAUTOTRAN;
      } else if (!strcmp(argv[i], "-oas")) {
        oflags |= KCOAUTOSYNC;
      } else if (!strcmp(argv[i], "-onl")) {
        oflags |= KCONOLOCK;
      } else if (!strcmp(argv[i], "-otl")) {
        oflags |= KCOTRYLOCK;
      } else if (!strcmp(argv[i], "-onr")) {
        oflags |= KCONOREPAIR;
      } else {
        usage();
      }
    } else if (!path) {
      argbrk = TRUE;
      path = argv[i];
    } else if (!rstr) {
      rstr = argv[i];
    } else {
      usage();
    }
  }
  if (!path || !rstr) usage();
  rnum = atoi(rstr);
  if (rnum < 1) usage();
  return procorder(path, rnum, rnd, etc, tran, oflags);
}


/* perform order command */
static int32_t procorder(const char* path, int64_t rnum, int32_t rnd, int32_t etc,
                         int32_t tran, int32_t oflags) {
  KCDB* db;
  KCCUR* cur, *paracur;
  int32_t err;
  char kbuf[RECBUFSIZ], *vbuf, wbuf[RECBUFSIZ], *corepath, *copypath, *snappath;
  size_t ksiz, vsiz, psiz;
  int32_t wsiz;
  int64_t i, cnt;
  double stime, etime;
  VISARG visarg;
  iprintf("<In-order Test>\n  path=%s  rnum=%ld  rnd=%d  etc=%d  tran=%d  oflags=%d\n\n",
          path, (long)rnum, rnd, etc, tran, oflags);
  err = FALSE;
  db = kcdbnew();
  iprintf("opening the database:\n");
  stime = kctime();
  if (!kcdbopen(db, path, KCOWRITER | KCOCREATE | KCOTRUNCATE | oflags)) {
    dberrprint(db, __LINE__, "kcdbopen");
    err = TRUE;
  }
  etime = kctime();
  dbmetaprint(db, FALSE);
  iprintf("time: %.3f\n", etime - stime);
  iprintf("setting records:\n");
  stime = kctime();
  for (i = 1; !err && i <= rnum; i++) {
    if (tran && !kcdbbegintran(db, FALSE)) {
      dberrprint(db, __LINE__, "kcdbbegintran");
      err = TRUE;
    }
    ksiz = sprintf(kbuf, "%08ld", (long)(rnd ? myrand(rnum) + 1 : i));
    if (!kcdbset(db, kbuf, ksiz, kbuf, ksiz)) {
      dberrprint(db, __LINE__, "kcdbset");
      err = TRUE;
    }
    if (tran && !kcdbendtran(db, TRUE)) {
      dberrprint(db, __LINE__, "kcdbendtran");
      err = TRUE;
    }
    if (rnum > 250 && i % (rnum / 250) == 0) {
      iputchar('.');
      if (i == rnum || i % (rnum / 10) == 0) iprintf(" (%08ld)\n", (long)i);
    }
  }
  etime = kctime();
  dbmetaprint(db, FALSE);
  iprintf("time: %.3f\n", etime - stime);
  if (etc) {
    iprintf("adding records:\n");
    stime = kctime();
    for (i = 1; !err && i <= rnum; i++) {
      if (tran && !kcdbbegintran(db, FALSE)) {
        dberrprint(db, __LINE__, "kcdbbegintran");
        err = TRUE;
      }
      ksiz = sprintf(kbuf, "%08ld", (long)(rnd ? myrand(rnum) + 1 : i));
      if (!kcdbadd(db, kbuf, ksiz, kbuf, ksiz) && kcdbecode(db) != KCEDUPREC) {
        dberrprint(db, __LINE__, "kcdbadd");
        err = TRUE;
      }
      if (tran && !kcdbendtran(db, TRUE)) {
        dberrprint(db, __LINE__, "kcdbendtran");
        err = TRUE;
      }
      if (rnum > 250 && i % (rnum / 250) == 0) {
        iputchar('.');
        if (i == rnum || i % (rnum / 10) == 0) iprintf(" (%08ld)\n", (long)i);
      }
    }
    etime = kctime();
    dbmetaprint(db, FALSE);
    iprintf("time: %.3f\n", etime - stime);
  }
  if (etc) {
    iprintf("appending records:\n");
    stime = kctime();
    for (i = 1; !err && i <= rnum; i++) {
      if (tran && !kcdbbegintran(db, FALSE)) {
        dberrprint(db, __LINE__, "kcdbbegintran");
        err = TRUE;
      }
      ksiz = sprintf(kbuf, "%08ld", (long)(rnd ? myrand(rnum) + 1 : i));
      if (!kcdbappend(db, kbuf, ksiz, kbuf, ksiz)) {
        dberrprint(db, __LINE__, "kcdbadd");
        err = TRUE;
      }
      if (tran && !kcdbendtran(db, TRUE)) {
        dberrprint(db, __LINE__, "kcdbendtran");
        err = TRUE;
      }
      if (rnum > 250 && i % (rnum / 250) == 0) {
        iputchar('.');
        if (i == rnum || i % (rnum / 10) == 0) iprintf(" (%08ld)\n", (long)i);
      }
    }
    etime = kctime();
    dbmetaprint(db, FALSE);
    iprintf("time: %.3f\n", etime - stime);
  }
  iprintf("getting records:\n");
  stime = kctime();
  for (i = 1; !err && i <= rnum; i++) {
    if (tran && !kcdbbegintran(db, FALSE)) {
      dberrprint(db, __LINE__, "kcdbbegintran");
      err = TRUE;
    }
    ksiz = sprintf(kbuf, "%08ld", (long)(rnd ? myrand(rnum) + 1 : i));
    vbuf = kcdbget(db, kbuf, ksiz, &vsiz);
    if (vbuf) {
      if (vsiz < ksiz || memcmp(vbuf, kbuf, ksiz)) {
        dberrprint(db, __LINE__, "kcdbget");
        err = TRUE;
      }
      kcfree(vbuf);
    } else if (!rnd || kcdbecode(db) != KCENOREC) {
      dberrprint(db, __LINE__, "kcdbget");
      err = TRUE;
    }
    if (tran && !kcdbendtran(db, TRUE)) {
      dberrprint(db, __LINE__, "kcdbendtran");
      err = TRUE;
    }
    if (rnum > 250 && i % (rnum / 250) == 0) {
      iputchar('.');
      if (i == rnum || i % (rnum / 10) == 0) iprintf(" (%08ld)\n", (long)i);
    }
  }
  etime = kctime();
  dbmetaprint(db, FALSE);
  iprintf("time: %.3f\n", etime - stime);
  if (etc) {
    iprintf("getting records with a buffer:\n");
    stime = kctime();
    for (i = 1; !err && i <= rnum; i++) {
      if (tran && !kcdbbegintran(db, FALSE)) {
        dberrprint(db, __LINE__, "kcdbbegintran");
        err = TRUE;
      }
      ksiz = sprintf(kbuf, "%08ld", (long)(rnd ? myrand(rnum) + 1 : i));
      wsiz = kcdbgetbuf(db, kbuf, ksiz, wbuf, sizeof(wbuf));
      if (wsiz >= 0) {
        if (wsiz < (int32_t)ksiz || memcmp(wbuf, kbuf, ksiz)) {
          dberrprint(db, __LINE__, "kcdbgetbuf");
          err = TRUE;
        }
      } else if (!rnd || kcdbecode(db) != KCENOREC) {
        dberrprint(db, __LINE__, "kcdbgetbuf");
        err = TRUE;
      }
      if (tran && !kcdbendtran(db, TRUE)) {
        dberrprint(db, __LINE__, "kcdbendtran");
        err = TRUE;
      }
      if (rnum > 250 && i % (rnum / 250) == 0) {
        iputchar('.');
        if (i == rnum || i % (rnum / 10) == 0) iprintf(" (%08ld)\n", (long)i);
      }
    }
    etime = kctime();
    dbmetaprint(db, FALSE);
    iprintf("time: %.3f\n", etime - stime);
  }
  if (etc) {
    iprintf("traversing the database by the inner iterator:\n");
    stime = kctime();
    cnt = kcdbcount(db);
    visarg.rnum = rnum;
    visarg.rnd = rnd;
    visarg.cnt = 0;
    memset(visarg.rbuf, '+', sizeof(visarg.rbuf));
    if (tran && !kcdbbegintran(db, FALSE)) {
      dberrprint(db, __LINE__, "kcdbbegintran");
      err = TRUE;
    }
    if (!kcdbiterate(db, visitfull, &visarg, TRUE)) {
      dberrprint(db, __LINE__, "kcdbiterate");
      err = TRUE;
    }
    if (rnd) iprintf(" (end)\n");
    if (tran && !kcdbendtran(db, TRUE)) {
      dberrprint(db, __LINE__, "kcdbendtran");
      err = TRUE;
    }
    if (visarg.cnt != cnt) {
      dberrprint(db, __LINE__, "kcdbiterate");
      err = TRUE;
    }
    etime = kctime();
    dbmetaprint(db, FALSE);
    iprintf("time: %.3f\n", etime - stime);
  }
  if (etc) {
    iprintf("traversing the database by the outer cursor:\n");
    stime = kctime();
    cnt = kcdbcount(db);
    visarg.rnum = rnum;
    visarg.rnd = rnd;
    visarg.cnt = 0;
    if (tran && !kcdbbegintran(db, FALSE)) {
      dberrprint(db, __LINE__, "kcdbbegintran");
      err = TRUE;
    }
    cur = kcdbcursor(db);
    if (!kccurjump(cur) && kccurecode(cur) != KCENOREC) {
      dberrprint(db, __LINE__, "kccurjump");
      err = TRUE;
    }
    paracur = kcdbcursor(db);
    while (!err && kccuraccept(cur, &visitfull, &visarg, TRUE, !rnd)) {
      if (rnd) {
        ksiz = sprintf(kbuf, "%08ld", (long)myrand(rnum));
        switch (myrand(3)) {
          case 0: {
            if (!kcdbremove(db, kbuf, ksiz) && kcdbecode(db) != KCENOREC) {
              dberrprint(db, __LINE__, "kcdbremove");
              err = TRUE;
            }
            break;
          }
          case 1: {
            if (!kccurjumpkey(paracur, kbuf, ksiz) && kccurecode(paracur) != KCENOREC) {
              dberrprint(db, __LINE__, "kccurjump");
              err = TRUE;
            }
            break;
          }
          default: {
            if (!kccurstep(cur) && kccurecode(cur) != KCENOREC) {
              dberrprint(db, __LINE__, "kccurstep");
              err = TRUE;
            }
            break;
          }
        }
      }
    }
    iprintf(" (end)\n");
    kccurdel(paracur);
    kccurdel(cur);
    if (tran && !kcdbendtran(db, TRUE)) {
      dberrprint(db, __LINE__, "kcdbendtran");
      err = TRUE;
    }
    if (!rnd && visarg.cnt != cnt) {
      dberrprint(db, __LINE__, "kccuraccept");
      err = TRUE;
    }
    etime = kctime();
    dbmetaprint(db, FALSE);
    iprintf("time: %.3f\n", etime - stime);
  }
  if (etc) {
    iprintf("synchronizing the database:\n");
    stime = kctime();
    if (!kcdbsync(db, FALSE, NULL, NULL)) {
      dberrprint(db, __LINE__, "kcdbsync");
      err = TRUE;
    }
    etime = kctime();
    dbmetaprint(db, FALSE);
    iprintf("time: %.3f\n", etime - stime);
  }
  if (etc) {
    corepath = kcdbpath(db);
    psiz = strlen(corepath);
    if (strstr(corepath, ".kch") || strstr(corepath, ".kct")) {
      copypath = kcmalloc(psiz + 256);
      sprintf(copypath, "%s.tmp", corepath);
      snappath = kcmalloc(psiz + 256);
      sprintf(snappath, "%s.kcss", corepath);
    } else {
      copypath = kcmalloc(256);
      sprintf(copypath, "kclangctest.tmp");
      snappath = kcmalloc(256);
      sprintf(snappath, "kclangctest.kcss");
    }
    iprintf("copying the database file:\n");
    stime = kctime();
    if (!kcdbcopy(db, copypath)) {
      dberrprint(db, __LINE__, "kcdbcopy");
      err = TRUE;
    }
    etime = kctime();
    dbmetaprint(db, FALSE);
    iprintf("time: %.3f\n", etime - stime);
    remove(copypath);
    iprintf("dumping records into snapshot:\n");
    stime = kctime();
    if (!kcdbdumpsnap(db, snappath)) {
      dberrprint(db, __LINE__, "kcdbdumpsnap");
      err = TRUE;
    }
    etime = kctime();
    dbmetaprint(db, FALSE);
    iprintf("time: %.3f\n", etime - stime);
    iprintf("loading records into snapshot:\n");
    stime = kctime();
    cnt = kcdbcount(db);
    if (rnd && myrand(2) == 0 && !kcdbclear(db)) {
      dberrprint(db, __LINE__, "kcdbclear");
      err = TRUE;
    }
    if (!kcdbloadsnap(db, snappath) || kcdbcount(db) != cnt) {
      dberrprint(db, __LINE__, "kcdbloadsnap");
      err = TRUE;
    }
    etime = kctime();
    dbmetaprint(db, FALSE);
    iprintf("time: %.3f\n", etime - stime);
    remove(snappath);
    kcfree(copypath);
    kcfree(snappath);
    kcfree(corepath);
  }
  iprintf("removing records:\n");
  stime = kctime();
  for (i = 1; !err && i <= rnum; i++) {
    if (tran && !kcdbbegintran(db, FALSE)) {
      dberrprint(db, __LINE__, "kcdbbegintran");
      err = TRUE;
    }
    ksiz = sprintf(kbuf, "%08ld", (long)(rnd ? myrand(rnum) + 1 : i));
    if (!kcdbremove(db, kbuf, ksiz) &&
        ((!rnd && !etc) || kcdbecode(db) != KCENOREC)) {
      dberrprint(db, __LINE__, "kcdbremove");
      err = TRUE;
    }
    if (tran && !kcdbendtran(db, TRUE)) {
      dberrprint(db, __LINE__, "kcdbendtran");
      err = TRUE;
    }
    if (rnum > 250 && i % (rnum / 250) == 0) {
      iputchar('.');
      if (i == rnum || i % (rnum / 10) == 0) iprintf(" (%08ld)\n", (long)i);
    }
  }
  etime = kctime();
  dbmetaprint(db, TRUE);
  iprintf("time: %.3f\n", etime - stime);
  iprintf("closing the database:\n");
  stime = kctime();
  if (!kcdbclose(db)) {
    dberrprint(db, __LINE__, "kcdbclose");
    err = TRUE;
  }
  etime = kctime();
  iprintf("time: %.3f\n", etime - stime);
  iprintf("%s\n\n", err ? "error" : "ok");
  kcdbdel(db);
  return err ? 1 : 0;
}



/* END OF FILE */
